sgkit_vcf/vcf_reader.py

```diff
@@ -0,0 +1,286 @@
+import itertools
+import shutil
+import tempfile
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Iterator, Optional, Sequence, Union
+
+import dask
+import numpy as np
+import xarray as xr
+from cyvcf2 import VCF, Variant
+
+from sgkit.api import DIM_VARIANT, create_genotype_call_dataset
+from sgkit.typing import PathType
+from sgkit_vcf.utils import chunks
+
+DEFAULT_ALT_NUMBER = 3  # see vcf_read.py in scikit_allel
+
+
+@contextmanager
+def open_vcf(path: PathType) -> VCF:
+    """A context manager for opening a VCF file."""
+    vcf = VCF(path)
+    try:
+        yield vcf
+    finally:
+        vcf.close()
+
+
+def region_filter(
+    variants: Iterator[Variant], region: Optional[str] = None
+) -> Iterator[Variant]:
+    """Filter out variants that don't start in the given region."""
+    if region is None:
+        return variants
+    else:
+        start = get_region_start(region)
+        return itertools.filterfalse(lambda v: v.POS < start, variants)
+
+
+def get_region_start(region: str) -> int:
+    """Return the start position of the region string."""
+    if ":" not in region:
+        return 1
+    contig, start_end = region.split(":")
+    start, end = start_end.split("-")
+    return int(start)
+
+
+def vcf_to_zarr_sequential(
+    input: PathType,
+    output: PathType,
+    region: Optional[str] = None,
+    chunk_length: int = 10_000,
+    chunk_width: int = 1_000,
+) -> None:
+
+    with open_vcf(input) as vcf:
+
+        alt_number = DEFAULT_ALT_NUMBER
+
+        sample_id = np.array(vcf.samples, dtype=str)
+        n_sample = len(sample_id)
+        n_allele = alt_number + 1
+        n_ploidy = 2
+
+        variant_contig_names = vcf.seqnames
+
+        # Remember max lengths of variable-length strings
+        max_variant_id_length = 0
+        max_variant_allele_length = 0
+
+        # Iterate through variants in batches of chunk_length
+
+        if region is None:
+            variants = vcf
+        else:
+            variants = vcf(region)
+
+        variant_contig = np.empty(chunk_length, dtype="i1")
+        variant_position = np.empty(chunk_length, dtype="i4")
+        call_genotype = np.empty((chunk_length, n_sample, n_ploidy), dtype="i1")
+        call_genotype_phased = np.empty((chunk_length, n_sample), dtype=bool)
```

Comment:
Assuming I'm understanding this correctly, I think it would be a good idea to have a `temp_chunk_length` argument in `vcf_to_zarr` that controls the number of rows here rather than using the target `chunk_length` in the final results.  Otherwise it will be very hard to fit all the sample chunks for one row of variant chunks in memory.

Putting a little more thought into this, the "auto" chunk size from dask for genome-wide UKB data is 11250x10000 so using 10k as an approximate chunk length (perhaps how you got the default in the first place?), a 10k x 500k array here would be problematic.

Reply:
I started doing this, but it needs a bit of investigation to get it working properly with the xarray API. I've opened https://github.com/pystatgen/sgkit-vcf/issues/2 to address later. (The workaround for the moment is just to use the smaller chunk length.)